s2/s2.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2022 Mostyn Bramley-Moore.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package s2 is an experimental wrapper for using
+// github.com/klauspost/compress/s2 stream compression with gRPC.
+package s2
+
+import (
+	"io"
+	"io/ioutil"
+	"sync"
+
+	"github.com/klauspost/compress/s2"
+	"google.golang.org/grpc/encoding"
+)
+
+const Name = "s2"
+
+type compressor struct {
+	poolCompressor   sync.Pool
+	poolDecompressor sync.Pool
+}
+
+type writer struct {
+	*s2.Writer
+	pool *sync.Pool
+}
+
+type reader struct {
+	*s2.Reader
+	pool *sync.Pool
+}
+
+func init() {
+	c := &compressor{}
+	c.poolCompressor.New = func() interface{} {
+		w := s2.NewWriter(ioutil.Discard, s2.WriterConcurrency(1))
```

Comment:
TODO: check if this is leaky. [s2.NewWriter](https://pkg.go.dev/github.com/klauspost/compress/s2#NewWriter) docs say that resources might not be released if Close isn't called (as would happen with sync.Pool):

> Users must call Close to guarantee all data has been forwarded to the underlying io.Writer and that resources are released. They may also call Flush zero or more times before calling Close.